### PR TITLE
frieren: restore volume-loss-weight=2.0 on SOTA stack (vol_pressure gap)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

BASELINE.md explicitly flags that PR #115 (current SOTA, test_abupt=10.580) did **NOT** include `--volume-loss-weight 2.0`. The note reads: *"The full SOTA stack should add it back — expected to recover volume_pressure and push abupt toward ~10.3."*

The volume_pressure axis has the largest gap to the AB-UPT reference (12.740 vs 6.08, ×2.1 gap), while surface_pressure is much closer to target (5.690 vs 3.82). Increasing the volume loss weight from the default (1.0) to 2.0 will push the optimizer to allocate more capacity to volume field predictions. Prior work (PR #141 askeladd, PR #142 thorfinn) confirmed that `--volume-loss-weight 2.0` improved abupt by ~10.78% in an earlier stack — it was subsequently dropped when building the current SOTA stack and never re-validated on Lion + EMA=0.999.

This is a **single-delta from the PR #115 SOTA**: the only change is adding `--volume-loss-weight 2.0`.

## Instructions

Run the following command exactly. The only change from the PR #115 SOTA is the addition of `--volume-loss-weight 2.0`:

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent frieren \
  --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --lion-beta1 0.9 --lion-beta2 0.99 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 \
  --volume-loss-weight 2.0 \
  --wandb-group tay-round12-vol-loss-weight-2p0
```

Report the best validation `abupt_axis_mean_rel_l2_pct` across all epochs, along with per-axis breakdown (surface_pressure, wall_shear, volume_pressure, tau_x, tau_y, tau_z) and the W&B run ID.

## Baseline

Current SOTA — PR #115 (thorfinn), W&B run `d03oghpp`:

| Metric | SOTA (PR #115) | AB-UPT Reference |
|--------|---------------|-----------------|
| **test_abupt** | **10.580** | — |
| val_abupt (ep9) | 9.484 | — |
| surface_pressure | 5.690 | 3.82 |
| wall_shear | 10.419 | 7.29 |
| volume_pressure | **12.740** | **6.08** |
| tau_x | 8.908 | — |
| tau_y | 12.491 | — |
| tau_z | 13.071 | — |

Expected outcome: val_abupt pushes toward ~9.2–9.4, volume_pressure drops significantly (target: ~10–11 range), test_abupt targets ~10.3.

Beat target: `val_abupt < 9.484` AND `volume_pressure < 12.740`

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: All 8 DDP ranks crashed at startup with zero training steps logged (W&B group tay-round12-vol-loss-weight-2p0 shows 20 failed runs across 2-3 attempts, no `_step` recorded on any rank). This is an infrastructure / pod-level crash, NOT a config error. Your config (volume_loss_weight=2.0 and the rest matching SOTA) is correct. Please restart the run. If it crashes again at startup, check pod logs for OOM / NCCL / data-loader errors and report what you find. Sending PR back to wip so you can resubmit when training is actually progressing.

**Comment by morganmcg1** [comment]
STUDENT frieren:

## Status update — original run is healthy and progressing

The original training launched at 12:43 UTC is still running successfully and has completed 6/9 epochs. The 16 failed runs you saw (named `frieren/vol-loss-weight-2p0-rank{0..7}` in the W&B group, started at 13:10 and 14:34) were duplicate launches that crashed with **CUDA OOM** because the original training was already occupying all 8 GPUs. The 8 active runs that ARE progressing are named `frieren/round12-vol-loss-weight-2p0-rank{0..7}` (note the `round12-` prefix) and have logged 16k+ steps each.

### Per-epoch validation progress (rank 0, run `qymdn7px`)

| Epoch | abupt | surf_p | vol_p | ws | ws_x | ws_y | ws_z |
|-------|-------|--------|-------|------|------|------|------|
| 1 | 55.509 | 40.483 | 37.846 | 60.096 | 50.573 | 77.389 | 71.252 |
| 2 | 25.452 | 18.155 | 13.707 | 28.495 | 24.183 | 34.084 | 37.130 |
| 3 | 16.941 | 11.981 | 8.745 | 19.083 | 16.112 | 22.979 | 24.886 |
| 4 | 13.654 | 9.552 | 7.262 | 15.357 | 13.048 | 18.106 | 20.302 |
| 5 | 11.907 | 8.054 | 6.557 | 13.437 | 11.431 | 15.999 | 17.492 |
| **6** | **10.969** | 7.301 | **6.187** | 12.385 | 10.556 | 14.757 | 16.045 |

### vs SOTA baseline (PR #115 ep9)

- **volume_pressure**: 6.19 vs 12.74 baseline — **already beating by 2.06×** at ep6
- val_abupt: 10.97 at ep6, slope -0.94/epoch → projected ~8.1 at ep9 (baseline ep9 was 9.48)
- surface_pressure: 7.30 (slightly above baseline 5.69 but improving)

3 more epochs to go (~90 min). I'll let it run to completion and report the final numbers. Resubmitting PR once we have the final epoch + test metrics.

**Comment by morganmcg1** [comment]
ADVISOR: W&B reading at step 17628 (ep~2-3): val_abupt=10.969%, slope -0.35pp/1k. Key signal: vol_pressure_mae=11.365 is BELOW SOTA 12.740 — the volume loss weighting is working as hypothesized. Wall_shear=12.385 is slightly above SOTA 10.632, but that may recover as training progresses. Continue to ep9 and report final metrics.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #204 because ADVISOR: Closing — vol_loss_weight=2.0 is a NEGATIVE result. Test abupt=11.096% vs SOTA 10.580% (+4.88%). Volume_pressure barely moved (12.772 vs 12.740, +0.25% ≈ neutral) but surface_pressure regressed -9.86% (5.690→6.251) and wall_shear -5.98% (10.419→11.042). Multi-task trade-off: boosting vol weight shifts capacity away from surface fidelity, which dominates the abupt aggregate. vol_loss_weight axis is now closed for this 9-epoch budget. Full review at https://github.com/wandb/senpai/pull/204#issuecomment-4360671523.

**Comment by morganmcg1** [comment]
STUDENT frieren:

## Results

**Verdict:** Hypothesis partially confirmed but **does NOT improve over PR #115 SOTA**. `--volume-loss-weight 2.0` dramatically reduced val volume_pressure (5.77% vs 12.74% baseline), but the optimizer reallocated capacity at the expense of every other axis, leaving **test_abupt 11.096 vs SOTA 10.580 (+4.9% regression)**. The val→test gap on volume is huge (vol_p_mae 10.53 → 28.76), so the val volume gain does not transfer.

### W&B run

- **Run ID:** `qymdn7px` ([rank 0 link](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/qymdn7px))
- **Group:** `tay-round12-vol-loss-weight-2p0`
- **Best epoch:** 9 (final), best by val abupt
- **Runtime:** 270.7 min, 23,534 global steps
- **Peak GPU memory:** 94.0 GiB / 95.6 GiB (98.3%) per H100, all 8 ranks

### Test metrics — `test_primary/*` (paper-facing)

| Metric | This run (vol_w=2.0) | SOTA PR #115 | Δ vs SOTA | AB-UPT ref |
|---|---:|---:|---:|---:|
| **abupt** | **11.096** | **10.580** | **+0.515 (+4.9%)** ❌ | — |
| surface_pressure | 6.251 | 5.690 | +0.561 (+9.9%) ❌ | 3.82 |
| wall_shear | 11.042 | 10.419 | +0.623 (+6.0%) ❌ | 7.29 |
| volume_pressure | 12.772 | 12.740 | +0.032 (+0.3%) ≈ | 6.08 |
| tau_x | 9.497 | 8.908 | +0.589 (+6.6%) ❌ | 5.35 |
| tau_y | 13.132 | 12.491 | +0.641 (+5.1%) ❌ | 3.65 |
| tau_z | 13.825 | 13.071 | +0.754 (+5.8%) ❌ | 3.63 |

### Validation metrics — `val_primary/*` (best/ep9)

| Metric | This run | SOTA PR #115 (val ep9) | Δ |
|---|---:|---:|---:|
| **abupt** | **9.945** | **9.484** | **+0.461 (+4.9%)** ❌ |
| surface_pressure | 6.524 | 5.690 (test) | +0.834 ❌ |
| wall_shear | 11.218 | 10.419 (test) | +0.799 ❌ |
| **volume_pressure** | **5.770** | **12.740 (test)** | **−6.970 (−54.7%)** ✅ |
| tau_x | 9.586 | 8.908 (test) | +0.678 ❌ |
| tau_y | 13.274 | 12.491 (test) | +0.783 ❌ |
| tau_z | 14.572 | 13.071 (test) | +1.501 ❌ |

(SOTA val per-axis ep9 not available in BASELINE.md, so val rows compare to SOTA test as an indicator.)

### Beat-target check (from PR body)

- ✅ `volume_pressure (val) < 12.740`: **5.770** (huge improvement)
- ❌ `val_abupt < 9.484`: **9.945** (worse by +0.461)

### Train.py command (from PR; ran exactly as instructed)

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent frieren \
  --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --lion-beta1 0.9 --lion-beta2 0.99 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999 \
  --volume-loss-weight 2.0 \
  --wandb-group tay-round12-vol-loss-weight-2p0
```

### What happened

The volume loss reweighting did exactly what was hypothesized to **on the val volume axis** — vol_pressure rel_l2 collapsed from ~12.7% to 5.77% (a 2.2× improvement, beating even AB-UPT's 6.08% reference!). However, this came with two unfortunate trade-offs:

1. **The other 5 axes regressed by 5–10% each.** The optimizer reallocated capacity from surface and wall_shear features to volume. Surface_pressure went from 5.69% → 6.25%, wall_shear from 10.42% → 11.04%, all three tau components regressed by ~6%. Because abupt is the mean over axes, a 1 axis improvement of −55% can't compensate for 5 axes of +6%, and abupt rose from 10.58 → 11.10.

2. **The val→test generalization gap on volume is enormous.** Val vol_p_mae = 10.53 but test vol_p_mae = 28.76 (×2.7 gap). Even though val_volume_pressure rel_l2 looks beautiful at 5.77%, **test_volume_pressure rel_l2 is 12.77% — essentially identical to the SOTA's 12.74%.** The volume reweighting helped the model fit the 34 val cases but did NOT extract a genuinely better-generalizing volume representation. Looking at the volume_pressure_mae trajectory, it keeps decreasing through all 9 epochs (slope −0.042/1k steps), suggesting more training would help, but the test gap suggests the geometry coverage of the val set vs test set is the dominant factor for volume.

The historical context says PR #141/#142 saw a +10.78% abupt improvement from `vol_w=2.0` on an earlier (Lion lr=5e-5) stack. That improvement does NOT replicate on the current Lion lr=1e-4 + EMA=0.999 stack — likely because the higher LR and stronger EMA already sharpen the optimizer's allocation across axes, so doubling volume's weight just over-tilts the balance instead of correcting an imbalance.

### Suggested follow-ups

1. **Try a lower volume weight** — `--volume-loss-weight 1.5` may capture some of the volume gain without pushing the surface/wall_shear axes off-balance. The current SOTA stack with `vol_w=1.0` may already be near-optimal balance for this LR/EMA combination.
2. **Investigate the val/test geometry mismatch on volume.** A 2.7× gap on vol_p_mae between val and test is a large generalization signal; the val set may not be representative for volume cases. Worth checking if a different val split or val-set subsampling exposes the issue.
3. **Try volume-specific regularization rather than just reweighting** — e.g. dropout on volume tokens, or a separate volume head with its own optimizer LR. This decouples capacity allocation from total loss balance.
4. **Decode why test volume is so much harder than val volume**. If 28.76 MAE on test is structural (e.g. specific test cases that are out-of-distribution), no reweighting will fix it without addressing data/geometry coverage.
